src/main/java/com/udojava/evalex/Expression.java

```diff
@@ -1782,6 +1788,51 @@ public String toRPN() {
 		return result.toString();
 	}
 
+	private void evalValidation() throws Exception {
+
+		final Set<String> circularcheck = new HashSet<String>();
+		try {
+			toRPN(circularcheck);
+		} catch (Exception e) {
+			throw e;
+		}
+	}
+
+	/**
+	 * circularcheck, throw Exception if circular definition
```

Comment:
That's a not very good method documentation. It should rather describe what the function does, and also note that it does a circular reference check. Or the method has the wrong name.

Reply:
I take your advice and sweep out early functions, and create a new varDefValidation function.
If expression has undefined vars or circular reference vars throw ExpressionException

Reply:
Some cases are wrong. Let me handle them

Reply:
I have fixed the cases, you can review it.